ibis/common/graph.py

```diff
@@ -4,100 +4,302 @@
+        for key, value in node.items():
+            yield from _flatten_collections(key, filter)
+            yield from _flatten_collections(value, filter)
+
+
+def _recursive_get(obj: Any, dct: dict[Node, Any]) -> Any:
+    """Recursively replace objects in a nested structure with values from a dict.
+
+    Since we treat collection types inherently traversable (e.g. list, tuple, dict) we
+    need to traverse them implicitly and replace the values given a result mapping.
+
+    Parameters
+    ----------
+    obj : Any
+        Object to replace.
+    dct : dict[Node, Any]
+        Mapping of objects to replace with their values.
+
+    Returns
+    -------
+    Object with replaced values.
+    """
+    if isinstance(obj, tuple):
+        return tuple(_recursive_get(o, dct) for o in obj)
+    elif isinstance(obj, dict):
+        return {k: _recursive_get(v, dct) for k, v in obj.items()}
+    else:
+        return dct.get(obj, obj)
 
 
 class Node(Hashable):
     __slots__ = ()
 
     @property
     @abstractmethod
-    def __args__(self) -> Sequence:
-        ...
+    def __args__(self) -> tuple[Any, ...]:
+        """Sequence of arguments to traverse."""
 
     @property
     @abstractmethod
-    def __argnames__(self) -> Sequence:
-        ...
-
-    def __children__(self, filter=None):
+    def __argnames__(self) -> tuple[str, ...]:
+        """Sequence of argument names."""
+
+    def __children__(self, filter: Optional[type] = None) -> tuple[Node, ...]:
+        """Return the children of this node.
+
+        This method is used to traverse the Node so it returns the children of the node
+        in the order they should be traversed. We treat common collection types
+        inherently traversable (e.g. list, tuple, dict), so this method flattens and
+        optionally filters the arguments of the node.
+
+        Parameters
+        ----------
+        filter : type, default Node
+            Type to filter out for the traversal, Node is used by default.
+
+        Returns
+        -------
+        Child nodes of this node.
+        """
         return tuple(_flatten_collections(self.__args__, filter or Node))
 
     def __rich_repr__(self):
+        """Support for rich reprerentation of the node."""
         return zip(self.__argnames__, self.__args__)
 
-    def map(self, fn, filter=None):
+    @experimental
+    def branches(self) -> Iterator[Sequence[Node]]:
+        """Yield all branches of the graph.
+
+        A branch is a path from the root to a leaf node. This method is primarily
+        used to implement the `path` method supporting `XPath`-like queries.
+
+        Yields
+        ------
+        A sequence of nodes representing a branch.
+        """
+        stack = [(self, [])]
+
+        while stack:
+            node, path = stack.pop()
+
+            if children := node.__children__():
+                for child in reversed(children):
+                    stack.append((child, path + [node]))
```

Comment:
This will not traverse the graph in topological order. Is that intentional?

Reply:
In this case the topological order is not relevant because we don't rewrite the tree, it just yields all branches and `node.path()` matches on those branches, but doesn't reconstruct the tree afterwards.

I'm not settled with this implementation yet, perhaps we won't need the `XPath`-like matching at all, then we can remove it.

Reply:
I think it might be relevant if there's a parent pattern that would match had it been traversed before a child pattern.

Consider this traversal order:

```mermaid
flowchart TD
    A[1]
    A --> B[2]
    A --> C[4]
    B --> D[3]
    C --> D  
```

Are there any issues with matching on 3 before matching on 4, if both match?

Reply:
For the following node: `A(B(C, D), E)`

```py
branches() ->
    [A, E]
    [A, B, D]
    [A, B, C]
```

`path()` iterates over that list of lists and tries to sequence match them.

So no pattern matching is happening during the traversal.

Reply:
This is totally experimental, we may not need it eventually, or we may come up with a better way to do matching like this - I kinda have an alternative idea.

Reply:
After discussing it, the problem comes from the ambiguity of the branch order, because `.path()` returns with the first match. I'm removing the `XPath` like matching until we actually need it. It will be in a separate commit so we can easily restore it.